CASSANDRA-17690 Instana case study (#139)
apache/cassandra-website: site-content/source/modules/ROOT/pages/case-studies/Instana.adoc

= Instana
:page-layout: case-study
:page-role: case-study

> :description: The Apache Cassandra Community

Reviewer: ```suggestion
:description: Instana queries billions of metrics with Apache Cassandra
```

:keywords: instana, apache cassandra

image::companies/instana_full.png[Instana,align="center"]

== Instana queries billions of metrics with Apache Cassandra

**Partner information**

> Observability & metrics

Reviewer: ```suggestion
* Observability & metrics
```

> 150+ employees

Reviewer: ```suggestion
* 150+ employees
```

> Millions of metrics per second 

Reviewer: ```suggestion
* Millions of metrics per second 
```

> Five production regions, each with 15 nodes

Reviewer: ```suggestion
* Five production regions, each with 15 nodes
```

> Running on both AWS and GCP

Reviewer: ```suggestion
* Running on both AWS and GCP
```

=== Graceful Degradation

Instana faced data-modeling challenges that monolithic architecture could not solve. Most notably, the application monitored hundreds of technologies and thousands of libraries, which users mixed into millions of configurations. Only a microservice architecture, supported by Apache Cassandra, could allow for graceful degradation when a single part of the data-receiving system stopped working or was deprecated.

=== Time Window Compaction

Early on, a big challenge was finding the right window and other settings for compactions. Without the right strategy, this could quickly run out of control. The compaction process merges keys, combines columns, evicts tombstones, consolidates SSTables, and creates a new index in the merged SSTable. Using this approach naively can result in millions of inserts per second when writing the new rolled-up data points. At one point, a single cluster had a single table with 30,000 SSTables, and the team had to run manual compactions to reduce it to a manageable size. 

> The scale of single-second data resolution did not make sense for long-term storage or queries that cover significant periods (e.g., looking at memory usage over the last three months). For this use-case, Instana wanted roll-up data, which indicates how data averaged over five-second intervals, for example. The choice was made to fix time window views for users, showing one-second, one-hour, one-day, and one-week views. Roll-up data has a different retention policy, and https://cassandra.apache.org/doc/trunk/cassandra/operating/compaction/index.html#ttl[Cassandra TTL] (Time to Live) was a natural fit for the data model.

Reviewer: ```suggestion
The scale of single-second data resolution did not make sense for long-term storage or queries that cover significant periods (e.g., looking at memory usage over the last three months). For this use-case, Instana wanted roll-up data, which indicates how data averaged over five-second intervals, for example. The choice was made to fix time window views for users, showing one-second, one-hour, one-day, and one-week views. Roll-up data has a different retention policy, and link:/doc/latest/cassandra/operating/compaction/index.adoc#ttl[Cassandra TTL] (Time to Live) was a natural fit for the data model.
```
